spacy_experimental/span_finder/span_finder_component.py

```diff
@@ -0,0 +1,352 @@
+        sgd: Optional[Optimizer] = None,
+        losses: Optional[Dict[str, float]] = None,
+    ) -> Dict[str, float]:
+        """Learn from a batch of documents and gold-standard information,
+        updating the pipe's model. Delegates to predict and get_loss.
+        examples (Iterable[Example]): A batch of Example objects.
+        drop (float): The dropout rate.
+        sgd (Optional[thinc.api.Optimizer]): The optimizer.
+        losses (Optional[Dict[str, float]]): Optional record of the loss during training.
+            Updated using the component name as the key.
+        RETURNS (Dict[str, float]): The updated losses dictionary.
+        """
+        if losses is None:
+            losses = {}
+        losses.setdefault(self.name, 0.0)
+        predicted = [eg.predicted for eg in examples]
+        set_dropout_rate(self.model, drop)
+        scores, backprop_scores = self.model.begin_update(predicted)
+        loss, d_scores = self.get_loss(examples, scores)
+        backprop_scores(d_scores)
+        if sgd is not None:
+            self.finish_update(sgd)
+        losses[self.name] += loss
+        return losses
+
+    def get_loss(self, examples, scores) -> Tuple[float, float]:
+        """Find the loss and gradient of loss for the batch of documents and
+        their predicted scores.
+        examples (Iterable[Examples]): The batch of examples.
+        scores: Scores representing the model's predictions.
+        RETURNS (Tuple[float, float]): The loss and the gradient.
+        """
+        reference_results = self._get_aligned_scores(examples)
+        d_scores = scores - self.model.ops.asarray(reference_results, dtype=float32)
+        loss = float((d_scores**2).sum())
+        return loss, d_scores
+
+    def _get_aligned_scores(self, examples) -> Tuple[Floats2d, Floats2d]:
+        """Align scores of the predictions to the references for calculating the loss"""
+        # TODO: handle 1-to-many alignments (see xfailed tests)
+        reference_results = []
+
+        for eg in examples:
+            start_indices = set()
+            end_indices = set()
+
+            if self.reference_key in eg.reference.spans:
+                for span in eg.reference.spans[self.reference_key]:
+                    start_indices.add(eg.reference[span.start].idx)
+                    end_indices.add(
+                        eg.reference[span.end - 1].idx + len(eg.reference[span.end - 1])
+                    )
+
+            for token in eg.predicted:
+                reference_start_score = 0
+                reference_end_score = 0
+                if token.idx in start_indices:
+                    reference_start_score = 1
+                if token.idx + len(token) in end_indices:
+                    reference_end_score = 1
+
+                reference_results.append(
+                    (
+                        reference_start_score,
+                        reference_end_score,
+                    )
+                )
```

Comment:
I think you can simplify this to:

```suggestion
            for token in eg.predicted:
                reference_results.append(
                    (
                        1 if token.idx in start_indices else 0,
                        1 if token.idx + len(token) in end_indices else 0,
                    )
                )
```